### PR TITLE
Fetch organisation list

### LINKED ISSUE #86: Retrieve list of organisations from real data set
### Description
We need to categorise the organisations.
Retrieve list of all organisations from REAL data, such that they can be organised. (issue #73)

#### Comments, Questions and Considerations
Include any relevant information

### Acceptance Criteria
This story can be considered done when:

**GIVEN**

**WHEN** 

**THEN** 
We have a list of all organisations existing in our data

### PR DESCRIPTION
**Pull request by Greatlemer**
Each **Pull Request** should relate to a single feature, therefore a single **Issue**. The smaller the **Pull Request** the faster it can be reviewed & merged. It should not be merged by the author.

## Issues to close

Fixes #86

## Notes

This is just a rough script that scans all the available csv files for what might be in the organisation column.
Comments for improvements very much welcome.

### CODE REVIEW

File: src/python/org_fetcher/fetch_orgs.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+PATH_TO_DATAFILES = os.path.join(
+    '..', '..', 'php', 'gov-harvester', 'datafiles'
+)
+
+
+orgs = defaultdict(lambda: 0)
+for filename in os.listdir(PATH_TO_DATAFILES):
+    filename_lower = filename.lower()
+    if not filename_lower.endswith('.csv'):
```
Comment (felipesere) [inline src/python/org_fetcher/fetch_orgs.py]: I think we could extract a method `is_csv(filename)` to make the intent a bit more obvious and potentially make the following comment obsolete.

File: src/python/org_fetcher/fetch_orgs.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        # Skip the file if it's not a csv
+        continue
+    if not 'meetings' in filename_lower:
```
Comment (felipesere) [inline src/python/org_fetcher/fetch_orgs.py]: Similar as above 😄 Maybe `contains_meetings(filename_lower)`?

File: src/python/org_fetcher/fetch_orgs.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        # Skip the file if it doesn't contain a meeting of some sort
+        continue
+    if 'gift' in filename_lower:
```
Comment (felipesere) [inline src/python/org_fetcher/fetch_orgs.py]: Same as above 😄 Maybe `contains_gift(filename_lower)`?

Come to think of it, we could take this a step further and a function called `should_skip(filename)` that just calls the three little functions from above:
```
function should_skip($filename) {
  return is_csv($filename) || contains_meetings($filename) || contains_gift($filename)
}
```

That method explains when we skip entire files 😸

File: src/python/org_fetcher/fetch_orgs.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        reader = csv.reader(file_handle)
+        first_row = True
+        org_col = None
+        for row in reader:
+            if len(row) < 3:
+                # If there's fewer than three columns this row is unlikely to be useful
+                # Mostly we'd want 4 but sometimes it's meetings for XXX followed by a
+                # three column version with the other details.
+                continue
+            if first_row and not row[1]:
```
Comment (felipesere) [inline src/python/org_fetcher/fetch_orgs.py]: I'd propose a similar pattern as above with `should_skip`. The comments can then be either removed or moved down to that function, to keep the main algorithm neat and tidy 😺

File: src/python/org_fetcher/fetch_orgs.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                # If the first row isn't full of headings then skip it and treat the
+                # next as if it's a heading
+                continue
+            if first_row:
+                # When looking at the heading row, scan across to find the column with
+                # the organisations in.
+                first_row = False
+                for index, col in enumerate(row):
```
Comment (felipesere) [inline src/python/org_fetcher/fetch_orgs.py]: We could extract this for loop into something like `find_organisation(row)`.
Make sure to return `None` to maintain the original flow of the code.

File: src/python/org_fetcher/fetch_orgs.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                    if 'Health' in col:
+                        # Special case, there's one file where this is in the minister
+                        # heading too.
+                        continue
+                    if 'organisation' in col.lower() or 'organization' in col.lower():
```
Comment (felipesere) [inline src/python/org_fetcher/fetch_orgs.py]: 👍 for catching American and British spelling!!!

**Review by felipesere — CHANGES_REQUESTED** [review]
Thanks for this PR!

I have left some comments around extracting functions to further surface the intent of the code.

**Comment by Greatlemer** [comment]
@felipesere I've updated my pull request with a lot more code extracted into bitesize functions.

I guess I was being a bit lazy with this one earlier as it was intended more as a one-shot script but you never know when someone else will want to modify and re-use it I suppose.

**Review by felipesere — APPROVED** [review]
Looks great!

**Review by felipesere — APPROVED** [review]
Looks great!

**Comment by felipesere** [comment]
How does the whole 'review required' thing work? @eddiejaoude @elischutze

**Comment by eddiejaoude** [comment]
You need someone else to `Accept` the **Pull Request**, then you or they can merge it :)

**Comment by eddiejaoude** [comment]
https://help.github.com/articles/about-pull-request-reviews/

**Comment by Greatlemer** [comment]
@eddiejaoude, Felipe was the reviewer, I am the author, is there a reason why he wouldn't have been able to approve it?

**Comment by eddiejaoude** [comment]
@Greatlemer oh i see.

@felipesere you need to accept the invite to the **HackBrexit** organisation. Then you will be able to approve the pull request & merge it.

screenshot 2016-11-25 07 55 19

**Review by elischutze — APPROVED** [review]

**Comment by elischutze** [comment]
Looks like you guys have this sorted so I'm going to go ahead and approve it. 👍

**Pull request merged**